Explorer/Assets/DCL/Multiplayer/Connections/Archipelago/Rooms/RealmRoomsProvider.cs

```diff
@@ -0,0 +1,86 @@
+        private readonly ICharacterObject characterObject;
+        private readonly IWebRequestController webRequestController;
+        private readonly IArchipelagoSignFlow archipelagoSignFlow;
+        private readonly ICurrentAdapterAddress currentAdapterAddress;
+
+        /// <summary>
+        /// Maintain a single instance to be able to reconfigure and share with all consumers
+        /// </summary>
+        private readonly InteriorRoom sharedRoom = new ();
+
+        private IRealmRoomStrategy? currentStrategy;
+
+        public RealmRoomsProvider(
+            IWeb3IdentityCache identityCache,
+            ICharacterObject characterObject,
+            IWebRequestController webRequestController,
+            IArchipelagoSignFlow archipelagoSignFlow,
+            ICurrentAdapterAddress currentAdapterAddress)
+        {
+            this.identityCache = identityCache;
+            this.characterObject = characterObject;
+            this.webRequestController = webRequestController;
+            this.archipelagoSignFlow = archipelagoSignFlow;
+            this.currentAdapterAddress = currentAdapterAddress;
+        }
+
+        public IRoom Room() =>
+            sharedRoom;
+
+        public IConnectiveRoom.State CurrentState() =>
+            currentStrategy?.ConnectiveRoom.CurrentState() ?? IConnectiveRoom.State.Stopped;
+
+        /// <summary>
+        /// Should be called from RealmController
+        /// </summary>
+        /// <param name="ct"></param>
+        /// <exception cref="InvalidOperationException"></exception>
+        public async UniTask StartAsync(CancellationToken ct)
+        {
+            if (currentStrategy != null && currentStrategy.ConnectiveRoom.CurrentState() is not IConnectiveRoom.State.Stopped)
+                throw new InvalidOperationException("First stop previous room before starting a new one");
+
+            string adapterUrl = await currentAdapterAddress.AdapterUrlAsync(ct);
+
+            if (adapterUrl.Contains("wss://"))
+                currentStrategy = CreateFixedConnectionRoomStrategy(adapterUrl);
+            else if (adapterUrl.Contains("https://"))
+                currentStrategy = CreateArchipelagoIslandRoomStrategy(adapterUrl);
+            else
+                throw new InvalidOperationException($"Cannot determine the protocol from the about url: {adapterUrl}");
+
+            currentStrategy.ConnectiveRoom.Start();
+        }
+
+        public UniTask StopAsync(CancellationToken ct) =>
+            currentStrategy?.ConnectiveRoom.StopAsync(ct) ?? UniTask.CompletedTask;
+
+        private FixedConnectionRoomStrategy CreateFixedConnectionRoomStrategy(string adapterUrl) =>
+            new (sharedRoom, webRequestController, adapterUrl);
+
+        private ArchipelagoIslandRoomStrategy CreateArchipelagoIslandRoomStrategy(string adapterUrl) =>
+            new (sharedRoom, identityCache, archipelagoSignFlow, characterObject, adapterUrl);
```

Comment:
Sharing an InteriorRoom can bring possibilities of mutating from different places that can be hard to debug

Why don't isolate it per each instanse?

Reply:
- According to our application it's not needed to be isolated in every instance
- Having it per instance created the problem with multi-leveled proxies when inside proxy there is another proxy, and so on and so forth: I described it in the PR

Reply:
> that can be hard to debug

Not sure if it is hard to debug: the source of `InteriorRoom` is streamlined and on purpose I no longer expose `InteriorRoom` from the core